### DISCUSSION TITLE
Discard submodule changes

### DISCUSSION THREAD

**Discussion by verhovsky** [magit/magit#4957] [Ideas]
How to reproduce:

Create a git repo and make the magit repo (as an example) its submodule

```sh
cd /tmp
mkdir foo
cd foo
git init
git submodule add https://github.com/magit/magit.git
git add .
git commit -m 'add magit submodule'
cd magit
git checkout 386d6c03456f547c915b3e7369ea54bef917c673 # some old commit
```

Now if you run magit status you'll see this:

<img width="761" alt="Screenshot 2023-06-14 at 6 37 39 PM" src="https://github.com/magit/magit/assets/5687998/ca051cca-341f-4e78-b490-a40b6f40715d">

because we added a submodule, commited that it's commit should be `9007bcfdb520f32befd48f8bf8c228e2a065f9e4` then `cd`'d into the submodule and changed its commit to a previous commit.

This all makes sense, but if you move the cursor to the submodule, press "x" to discard the changes, it'll ask if you're sure and if you say "y" it won't do anything whereas I expect it to run `git checkout 9007bcfdb520f32befd48f8bf8c228e2a065f9e4` on the submodule.

One detail here is what to do if there are unstaged changes in the submodule, in that case I guess you need to throw an error.

**Comment by kyleam** [comment]
> This all makes sense, but if you move the cursor to the submodule, press "x" to discard the changes [...]

By default `x` is bound to `magit-reset-quickly`, but I assume you're talking about `magit-discard` (bound to `k` by default).

Related:

 * gh-134

 * [gh-2657](https://github.com/magit/magit/issues/2657#issuecomment-220852585)

   > > It's a bit annoying that `k` (`magit-discard`) does nothing on submodule sections.
   >
   > What should it do? This is actually one command I currently think should not be taught about submodules. Resetting to the recorded commit seems reasonable but it's also dangerous because you could end up losing commits. In any case I would not want this command to result in a detached `HEAD`.

**Reply by verhovsky** [reply]
Yea you're right, I'm using Doom Emacs which rebinds magit keys.

**Reply by verhovsky** [reply]
> What should it do? This is actually one command I currently think should not be taught about submodules. Resetting to the recorded commit seems reasonable but it's also dangerous because you could end up losing commits. In any case I would not want this command to result in a detached HEAD.

in my case I am 100% of the time not doing work on the submodules. If there was just 7 different checks to make sure that undoing it is safe and it only undid commits that it was certain it was safe I would find that quite valuable. It is frustrating that it asks for confirmation but doesn't do anything.

it should at least stop asking for confirmation.
